# Lab 1.3: Exploratory Data Analysis (EDA) with Pandas

## Outline

* Introduction to Pandas
* EDA using Pandas

Exploratory data analysis is a first crucial step to building predictive models from your data. EDA allows you
to confirm or invalidate some of the assumptions you are making about your data and understand relationships between your variables.

## Getting Started with [Pandas](http://pandas.pydata.org/pandas-docs/version/0.17.1/)

### What is Pandas?
A Python library providing data structures and data analysis tools.

### Huh?
Think of it like Excel for Python but better.

In [112]:
#By convention import pandas like:
import pandas as pd

In [113]:
#For fake data.
from numpy.random import randn
import numpy as np

## Series
Think of a Pandas Series as a _labeled_ one-dimensional vector. In fact, it need not be a numeric vector, it can contain arbitrary python objects.

In [114]:
int_series = pd.Series(range(10))
int_series.head()

0    0
1    1
2    2
3    3
4    4
dtype: int64

In [115]:
num_series = pd.Series(randn(10))
num_series.head()

0    0.486160
1    0.187037
2    0.084080
3   -0.228814
4    1.743414
dtype: float64

In [116]:
str_series = pd.Series([x for x in 'abcde'*2])
str_series.head()

0    a
1    b
2    c
3    d
4    e
dtype: object

## Indexes
Notice how each series has an index (in this case a relatively meaningless default index).

Pandas can make great use of informative indexes. Indexes work similarly to a dictionary key, allowing fast lookups of the data associated with the index.

Indexes can also be exploited for fast group-bys, merges, time-series operations and lots more.

When you're really in the zone with pandas, you'll be thinking a lot about indexes.

In [117]:
indexed_series = pd.Series(randn(5), 
                           index = ['California', 'Alabama', 
                                    'Indiana', 'Montana', 
                                    'Kentucky'])
alt_indexed_series = pd.Series(randn(5),
                               index = ['Washington', 'Alabama', 
                                        'Montana', 'Indiana', 
                                        'New York'])
print (indexed_series)
print ('\n')
print (alt_indexed_series)

California    0.193554
Alabama      -0.254618
Indiana       0.320663
Montana       1.798329
Kentucky     -1.852089
dtype: float64


Washington    1.031923
Alabama      -0.474125
Montana       0.118653
Indiana       0.218213
New York     -0.162685
dtype: float64


In [118]:
#Pandas uses the index by default to align series for arithmetic!
indexed_series + alt_indexed_series

Alabama      -0.728743
California         NaN
Indiana       0.538876
Kentucky           NaN
Montana       1.916982
New York           NaN
Washington         NaN
dtype: float64

In [119]:
indexed_series = pd.Series(randn(5*10), 
                           index = ['California', 'Alabama', 
                                    'Indiana', 'Montana', 
                                    'Kentucky']*10)
#If you have non-unique indexes, you can use them 
#to do groupby operations.
indexed_series.groupby(level=0).mean()

Alabama       0.167500
California   -0.121792
Indiana       0.176649
Kentucky     -0.376906
Montana       0.583118
dtype: float64

In [120]:
#Datetime index
dt_index = pd.date_range('2015-1-1', 
                        '2015-11-1', 
                        freq='m')
dt_series = pd.Series(randn(10), 
                      index = dt_index)
dt_series

2015-01-31   -0.596381
2015-02-28    0.251437
2015-03-31    0.279454
2015-04-30    0.990507
2015-05-31    0.033152
2015-06-30    0.277157
2015-07-31    1.662184
2015-08-31    0.533490
2015-09-30    0.925641
2015-10-31    0.306172
Freq: M, dtype: float64

In [121]:
#Datetime indexes make it easy to transform freqs etc!
dt_series.resample('q').mean()

2015-03-31   -0.021830
2015-06-30    0.433605
2015-09-30    1.040438
2015-12-31    0.306172
Freq: Q-DEC, dtype: float64

## DataFrames
Data frames extend the concept of Series to table-like data.

In [122]:
df = pd.DataFrame(randn(10, 5), index=dt_index, columns=[x for x in 'abcde'])
df

a         b         c         d         e
2015-01-31  0.182232  0.094585  0.316332 -1.002885  0.441167
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695
2015-03-31 -0.435876 -0.556900  0.956910 -0.266178  1.604977
2015-04-30  0.717300  0.812005 -1.152371  0.011826  0.378750
2015-05-31 -1.671123  0.801658 -1.413547 -1.940020 -0.701348
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658
2015-07-31 -1.069421  0.112683  0.085844  1.210465 -0.180127
2015-08-31 -1.846665 -0.733902 -3.021995  0.234681  0.067132
2015-09-30 -0.511256 -0.285234  0.775043  0.004805  0.344354
2015-10-31  0.116062  0.743007  0.118452  0.893905  0.942008

In [123]:
#A dataframes columns are series:
col = df.a
type(col)

pandas.core.series.Series

In [124]:
#So are the rows.
row = df.ix['2015-01-31']
type(row)

pandas.core.series.Series

In [125]:
#The columns all have the same index:
col.index   

DatetimeIndex(['2015-01-31', '2015-02-28', '2015-03-31', '2015-04-30',
               '2015-05-31', '2015-06-30', '2015-07-31', '2015-08-31',
               '2015-09-30', '2015-10-31'],
              dtype='datetime64[ns]', freq='M')

In [126]:
#What's the index for the rows?
row.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

## DataFrame basics

In [127]:
#New column
df['new'] = df['a'] + df['b']
df


a         b         c         d         e       new
2015-01-31  0.182232  0.094585  0.316332 -1.002885  0.441167  0.276818
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695  0.010712
2015-03-31 -0.435876 -0.556900  0.956910 -0.266178  1.604977 -0.992776
2015-04-30  0.717300  0.812005 -1.152371  0.011826  0.378750  1.529305
2015-05-31 -1.671123  0.801658 -1.413547 -1.940020 -0.701348 -0.869466
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658  2.040460
2015-07-31 -1.069421  0.112683  0.085844  1.210465 -0.180127 -0.956738
2015-08-31 -1.846665 -0.733902 -3.021995  0.234681  0.067132 -2.580567
2015-09-30 -0.511256 -0.285234  0.775043  0.004805  0.344354 -0.796490
2015-10-31  0.116062  0.743007  0.118452  0.893905  0.942008  0.859069

In [128]:
#Delete a column
df.drop('new', axis=1)  #axis=1 to look at columns, instead of rows.

# df.drop?

a         b         c         d         e
2015-01-31  0.182232  0.094585  0.316332 -1.002885  0.441167
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695
2015-03-31 -0.435876 -0.556900  0.956910 -0.266178  1.604977
2015-04-30  0.717300  0.812005 -1.152371  0.011826  0.378750
2015-05-31 -1.671123  0.801658 -1.413547 -1.940020 -0.701348
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658
2015-07-31 -1.069421  0.112683  0.085844  1.210465 -0.180127
2015-08-31 -1.846665 -0.733902 -3.021995  0.234681  0.067132
2015-09-30 -0.511256 -0.285234  0.775043  0.004805  0.344354
2015-10-31  0.116062  0.743007  0.118452  0.893905  0.942008

## Axis?
Because pandas thinks of rows and columns as both being series, anything we can do to rows we can do to columns too. 

Axis describes which one we want to do it to. 0=rows, 1=columns.

In [129]:
lbl = pd.Timestamp('2015-04-30 00:00:00', offset='M')
# pd.Timestamp?
df.drop(lbl, axis=0)

a         b         c         d         e       new
2015-01-31  0.182232  0.094585  0.316332 -1.002885  0.441167  0.276818
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695  0.010712
2015-03-31 -0.435876 -0.556900  0.956910 -0.266178  1.604977 -0.992776
2015-05-31 -1.671123  0.801658 -1.413547 -1.940020 -0.701348 -0.869466
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658  2.040460
2015-07-31 -1.069421  0.112683  0.085844  1.210465 -0.180127 -0.956738
2015-08-31 -1.846665 -0.733902 -3.021995  0.234681  0.067132 -2.580567
2015-09-30 -0.511256 -0.285234  0.775043  0.004805  0.344354 -0.796490
2015-10-31  0.116062  0.743007  0.118452  0.893905  0.942008  0.859069

## Selecting subsets
There a couple of ways to select a subset of a python data frame.

In [130]:
#To subset the rows, you can use the convenient:
df[df.a > 0]

a         b         c         d         e       new
2015-01-31  0.182232  0.094585  0.316332 -1.002885  0.441167  0.276818
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695  0.010712
2015-04-30  0.717300  0.812005 -1.152371  0.011826  0.378750  1.529305
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658  2.040460
2015-10-31  0.116062  0.743007  0.118452  0.893905  0.942008  0.859069

In [131]:
#Or combine multiple conditions:
df[(df.a > 0) & (df.c < 0)]

a         b         c         d         e       new
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695  0.010712
2015-04-30  0.717300  0.812005 -1.152371  0.011826  0.378750  1.529305

In [132]:
#Selecting a column
df.a
#Works sometimes.

2015-01-31    0.182232
2015-02-28    1.459722
2015-03-31   -0.435876
2015-04-30    0.717300
2015-05-31   -1.671123
2015-06-30    1.118365
2015-07-31   -1.069421
2015-08-31   -1.846665
2015-09-30   -0.511256
2015-10-31    0.116062
Freq: M, Name: a, dtype: float64

In [133]:
df['a'] #works always

2015-01-31    0.182232
2015-02-28    1.459722
2015-03-31   -0.435876
2015-04-30    0.717300
2015-05-31   -1.671123
2015-06-30    1.118365
2015-07-31   -1.069421
2015-08-31   -1.846665
2015-09-30   -0.511256
2015-10-31    0.116062
Freq: M, Name: a, dtype: float64

In [134]:
df[['a', 'b']]
#Or a subset of columns

a         b
2015-01-31  0.182232  0.094585
2015-02-28  1.459722 -1.449010
2015-03-31 -0.435876 -0.556900
2015-04-30  0.717300  0.812005
2015-05-31 -1.671123  0.801658
2015-06-30  1.118365  0.922095
2015-07-31 -1.069421  0.112683
2015-08-31 -1.846665 -0.733902
2015-09-30 -0.511256 -0.285234
2015-10-31  0.116062  0.743007

## Advanced selection
The above methods work for simple types of selections, but what if you need something more complex?

In [135]:
df.loc['2015-05-31':'2015-08-31', 'c':'e'] #Ranges by label.

c         d         e
2015-05-31 -1.413547 -1.940020 -0.701348
2015-06-30  0.267316  0.951993  0.917658
2015-07-31  0.085844  1.210465 -0.180127
2015-08-31 -3.021995  0.234681  0.067132

In [136]:
df.iloc[2:-3,2:5] #Ranges by number.

c         d         e
2015-03-31  0.956910 -0.266178  1.604977
2015-04-30 -1.152371  0.011826  0.378750
2015-05-31 -1.413547 -1.940020 -0.701348
2015-06-30  0.267316  0.951993  0.917658
2015-07-31  0.085844  1.210465 -0.180127

In [137]:
#SQL-like queries for parsimonious row selection.
#Experimental
df.query('a > 1 and c <1')

a         b         c         d         e       new
2015-02-28  1.459722 -1.449010 -0.147833 -0.546533  0.551695  0.010712
2015-06-30  1.118365  0.922095  0.267316  0.951993  0.917658  2.040460

In [147]:
#Multi Index:
dt_index = pd.date_range('2015-1-1', 
                        '2017-7-1', 
                        freq='m')
df = pd.DataFrame(randn(30,5), index=dt_index)

In [148]:
df['state'] = ['Alabama', 'Alaska' , 'Arizona'] * 10
df.head()

0         1         2         3         4    state
2015-01-31 -1.114151 -0.699261 -0.432288  0.294191  0.802589  Alabama
2015-02-28 -1.139904 -0.834198  1.425801  0.663338 -1.265726   Alaska
2015-03-31  2.622086 -0.106730  1.798027  0.819095 -0.082138  Arizona
2015-04-30 -0.373654  0.681551  0.134425 -0.662291  0.346853  Alabama
2015-05-31 -0.361792  1.213761 -0.395428  0.304106  0.366492   Alaska

In [149]:
df = df.reset_index()
df = df.set_index(['state', 'index'])
df.head()

0         1         2         3         4
state   index                                                       
Alabama 2015-01-31 -1.114151 -0.699261 -0.432288  0.294191  0.802589
Alaska  2015-02-28 -1.139904 -0.834198  1.425801  0.663338 -1.265726
Arizona 2015-03-31  2.622086 -0.106730  1.798027  0.819095 -0.082138
Alabama 2015-04-30 -0.373654  0.681551  0.134425 -0.662291  0.346853
Alaska  2015-05-31 -0.361792  1.213761 -0.395428  0.304106  0.366492

In [152]:
df.loc['Alabama'].head()

# df.loc?

In [153]:
df.loc['2015-01-31'] #Doesn't work.

KeyError: 'the label [2015-01-31] is not in the [index]'

In [156]:
df.loc[('Alabama', '2015-01-31')] #Can do this.

0   -1.114151
1   -0.699261
2   -0.432288
3    0.294191
4    0.802589
Name: (Alabama, 2015-01-31 00:00:00), dtype: float64

In [157]:
#Can also have multi-index in columns.
df.reset_index().set_index(['index', 'state']).unstack().head()

0                             1                     \
state        Alabama    Alaska   Arizona   Alabama    Alaska  Arizona   
index                                                                   
2015-01-31 -1.114151       NaN       NaN -0.699261       NaN      NaN   
2015-02-28       NaN -1.139904       NaN       NaN -0.834198      NaN   
2015-03-31       NaN       NaN  2.622086       NaN       NaN -0.10673   
2015-04-30 -0.373654       NaN       NaN  0.681551       NaN      NaN   
2015-05-31       NaN -0.361792       NaN       NaN  1.213761      NaN   

                   2                             3                      \
state        Alabama    Alaska   Arizona   Alabama    Alaska   Arizona   
index                                                                    
2015-01-31 -0.432288       NaN       NaN  0.294191       NaN       NaN   
2015-02-28       NaN  1.425801       NaN       NaN  0.663338       NaN   
2015-03-31       NaN       NaN  1.798027       NaN       NaN  0.819095   
2015-04-30  0.134425       NaN       NaN -0.662291       NaN       NaN   
2015-05-31       NaN -0.395428       NaN       NaN  0.304106       NaN   

                   4                      
state        Alabama    Alaska   Arizona  
index                                     
2015-01-31  0.802589       NaN       NaN  
2015-02-28       NaN -1.265726       NaN  
2015-03-31       NaN       NaN -0.082138  
2015-04-30  0.346853       NaN       NaN  
2015-05-31       NaN  0.366492       NaN

In [163]:
df.head()

0         1         2         3         4
state   index                                                       
Alabama 2015-01-31 -1.114151 -0.699261 -0.432288  0.294191  0.802589
Alaska  2015-02-28 -1.139904 -0.834198  1.425801  0.663338 -1.265726
Arizona 2015-03-31  2.622086 -0.106730  1.798027  0.819095 -0.082138
Alabama 2015-04-30 -0.373654  0.681551  0.134425 -0.662291  0.346853
Alaska  2015-05-31 -0.361792  1.213761 -0.395428  0.304106  0.366492

## split-apply-combine

In [165]:
df.groupby(level=1).mean().head()

0         1         2         3         4
index                                                       
2015-01-31 -1.114151 -0.699261 -0.432288  0.294191  0.802589
2015-02-28 -1.139904 -0.834198  1.425801  0.663338 -1.265726
2015-03-31  2.622086 -0.106730  1.798027  0.819095 -0.082138
2015-04-30 -0.373654  0.681551  0.134425 -0.662291  0.346853
2015-05-31 -0.361792  1.213761 -0.395428  0.304106  0.366492

In [166]:
df.groupby(level=0).mean().head() #Groupby index

0         1         2         3         4
state                                                    
Alabama -0.509177  0.349831 -0.248036 -0.173866  0.041430
Alaska  -0.628315  0.765471  0.260982  0.310699  0.300114
Arizona  0.461940  0.203121  0.025535 -0.233001 -0.343813

In [175]:
#Groupby doesn't have to be on index.
df.reset_index().groupby('state').mean().head()
#'state' could be a list of columns to group on.

0         1         2         3         4
state                                                    
Alabama -0.509177  0.349831 -0.248036 -0.173866  0.041430
Alaska  -0.628315  0.765471  0.260982  0.310699  0.300114
Arizona  0.461940  0.203121  0.025535 -0.233001 -0.343813

In [176]:
#Apply
g = df.groupby(level=0)

In [178]:
#If one row for each index.
g.aggregate(np.mean)

0         1         2         3         4
state                                                    
Alabama -0.509177  0.349831 -0.248036 -0.173866  0.041430
Alaska  -0.628315  0.765471  0.260982  0.310699  0.300114
Arizona  0.461940  0.203121  0.025535 -0.233001 -0.343813

In [182]:
#If same shape as original:
g.transform(lambda x: (x - x.mean())/x.std()).head()

0         1         2         3         4
state   index                                                       
Alabama 2015-01-31 -0.575780 -1.011110 -0.135385  0.808546  0.614796
Alaska  2015-02-28 -0.468890 -1.643689  1.416974  0.321334 -1.162098
Arizona 2015-03-31  2.160973 -0.353863  1.268149  1.321768  0.425305
Alabama 2015-04-30  0.128984  0.319710  0.281026 -0.843730  0.246693
Alaska  2015-05-31  0.244278  0.460626 -0.798507 -0.006007  0.049263

In [186]:
#Different index than I started with.
g.apply(lambda x: x.describe()).head(12)

ValueError: Buffer dtype mismatch, expected 'Python object' but got 'long'

ValueError: Buffer dtype mismatch, expected 'Python object' but got 'long'

## Reading and writing data

In [ ]:
df.to_csv('saved_data.csv', delimiter='\t')

In [ ]:
read = pd.read_csv('saved_data.csv', delimiter='\t')
#TONS OF options for reading data

"""Other methods:
pd.read_excel
pd.read_sql
pd.read_stata
...
"""

## Other very useful things

join, concat and merge

In [ ]:
state_df = pd.DataFrame({'governor':['Robert Bentley',
                                    'Bill Walker',
                                    'Doug Ducey',
                                    'Asa Hutchinson']}, 
                        index=['Alabama', 'Alaska', 'Arizona', 'Arkansas'])
#Note merge is most useful when you want to merge on something other than the index.
#Default is to merge on common column names.
pd.merge(df.reset_index(), state_df, 
         left_on='state', right_index=True, how='right').tail()

In [ ]:
state_avg = df.groupby(level=0).mean()
state_avg.head()

In [ ]:
#Concat allows joining along the axes.
pd.concat([state_avg, state_df], axis=1)

In [ ]:
#Join also works. Just like merge, but 
#default is to join on indexes.
state_avg.join(state_df, how='right')

## Exploratory Data Analysis with Pandas

In [ ]:
%pylab inline

In [ ]:
df = pd.read_csv('data/playgolf.csv', delimiter='|' )
print df.head()

## Describe the continuous variables
### This treats the Boolean Windy variable as a series of 0's and 1's

In [ ]:
df.describe()

Can see the general pattern of Temperature and Humidity and mean of a Boolean represents the percentage

### We can make use of df.plot() to produce simple graphs that calls on the more adjustable [Matplotlib](http://matplotlib.org/api/pyplot_api.html) library 

In [ ]:
df.hist(['Temperature','Humidity'],bins=5)

In [ ]:
df[['Temperature','Humidity']].plot(kind='box')

### Scatterplots for examining bivariate relationships

In [ ]:
df.plot('Temperature', 'Humidity', kind='scatter')

## What about the categorical variables? Frequency tables and relative frequency tables

### Simply df.value_counts() gets you the frequencies

In [ ]:
df['Outlook'].value_counts()

### Using apply will get you the value counts for multiple columns at once

In [ ]:
df[['Outlook','Result']].apply(lambda x: x.value_counts())

### Contingency Tables for looking at bivariate relationships between two categorical variables

In [ ]:
pd.crosstab(df['Outlook'], df['Result'])

### Often we want the row percentages

In [ ]:
pd.crosstab(df['Outlook'], df['Result']).apply(lambda r: r/r.sum(), axis=1)

### Or the column percentages

In [ ]:
pd.crosstab(df['Outlook'], df['Result']).apply(lambda c: c/c.sum(), axis=0)

## Lab Exercises

In this scenario, you are a data scientist at [Bay Area Bike Share](http://www.bayareabikeshare.com/). Your task
is to provide insights on bike user activity and behavior to the products team. 


1. Load the `trips` table into a dataframe. 
   
   Make 4 extra columns from the `start_date` column (We will use these in later questions):
   - `month` would contain only the month component
   - `dayofweek` would indicate what day of the week the date is
   - `date` would contain only the date component 
   - `hour` would only contain the hour component
   - [Hint to deal with datetime objects in pandas](http://stackoverflow.com/questions/25129144/pandas-return-hour-from-datetime-column-directly)

2. Group the bike rides by `month` and count the number of users per month. Plot the number of users for each month. 
   What do you observe? Provide a likely explanation to your observation. Real life data can often be messy/incomplete
   and cursory EDA is often able to reveal that.
   
3. Plot the daily user count from September to December. Mark the `mean` and `mean +/- 1.5 * Standard Deviation` as 
   horizontal lines on the plot. This would help you identify the outliers in your data. Describe your observations. 
   
   ![image](images/timeseries.png)

4. Plot the distribution of the daily user counts for all months as a histogram. Comment on the distribution you see and explain why the distribution might be shaped as such. 
    
   <img src="images/hist.png" width="500">
  
5. Now we are going to explore hourly trends of user activity. Group the bike rides by `date` and `hour` and count 
   the number of rides in the given hour on the given date. Make a 
   [boxplot](http://blog.bharatbhole.com/creating-boxplots-with-matplotlib/) of the hours in the day **(x)** against
   the number of users **(y)** in that given hour. 
   
6. Someone from the analytics team made a line plot (_right_) that he claims is showing the same information as your
   boxplot (_left_). What information can you gain from the boxplot that is missing in the line plot?
   
   ![image](images/q1_pair.png)

7. ** Extra Credit **: Replot the boxplot in `6.` after binning your data into weekday and weekend. Describe the differences you observe between hour user activity between weekday and weekend? 
    